changefeedccl: support updated and resolved timestamps in avro

danhhz opened the pull request:
To do this, the avro format is changed in a backward-incompatible way.
This is unfortunate but okay since we've marked avro as experimental
with exactly this sort of thing in mind.

We map a SQL table schema to an Avro record with 1:1 mapping between
table columns and Avro fields. The type of the column is mapped to a
native Avro type as faithfully as possible. This is then used to make an
"optional" Avro field for that column by unioning with null and
explictly specifying null as default, regardless of whether the sql
column allows NULLs. This may seem an odd choice, but it allows for all
adjacent Avro schemas for a given SQL table to be backward and forward
compatible with each other. Forward and backward compatibility
drastically eases use of the resulting data by downstream systems,
especially when working with long histories of archived data across many
schema changes (such as a data lake).

One downside of the above is that it's not possible to recover the
original SQL table schema from an Avro one. (This is also true for other
reasons, such as lossy mappings from sql types to avro types.) To
partially address this, the SQL column type is embedded as metadata in
the Avro field schema in a way that Avro ignores it but passes it along.

I have a roachest that verifies the compatiblity guarantees using the
Confluent Schema Registry, but it still needs some polish and I wanted
to get this part out now.

Release note (backward-incompatible change): `CHANGEFEED`'s
experimental avro format has changed. The new format is backward and
forward compatible with adjacent schemas for the same table.

danhhz committed:
changefeedccl: support updated and resolved timestamps in avro

To do this, the avro format is changed in a backward-incompatible way.
This is unfortunate but okay since we've marked avro as experimental
with exactly this sort of thing in mind.

We map a SQL table schema to an Avro record with 1:1 mapping between
table columns and Avro fields. The type of the column is mapped to a
native Avro type as faithfully as possible. This is then used to make an
"optional" Avro field for that column by unioning with null and
explictly specifying null as default, regardless of whether the sql
column allows NULLs. This may seem an odd choice, but it allows for all
adjacent Avro schemas for a given SQL table to be backward and forward
compatible with each other. Forward and backward compatibility
drastically eases use of the resulting data by downstream systems,
especially when working with long histories of archived data across many
schema changes (such as a data lake).

One downside of the above is that it's not possible to recover the
original SQL table schema from an Avro one. (This is also true for other
reasons, such as lossy mappings from sql types to avro types.) To
partially address this, the SQL column type is embedded as metadata in
the Avro field schema in a way that Avro ignores it but passes it along.

I have a roachest that verifies the compatiblity guarantees using the
Confluent Schema Registry, but it still needs some polish and I wanted
to get this part out now.

Release note (backward-incompatible change): `CHANGEFEED`'s
experimental avro format has changed. The new format is backward and
forward compatible with adjacent schemas for the same table.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34095)
<!-- Reviewable:end -->

danhhz committed:
workload: add a `FromFlags` helper for using workload in unit tests

Bank and TPCC already had something like this, but they didn't go
through the normal codepaths so the workload was sometimes left in a
wonky state. Plus this has been useful enough that it's worth a helper.

Release note: None

danhhz committed:
changefeedccl: add a test for the ledger workload with avro

Release note: None

mrtracy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34095)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @mrtracy)

___
*[pkg/ccl/changefeedccl/avro.go, line 510 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34095#-LWXYI6D0S_Ae2uBKMO4:-LWXYI6D0S_Ae2uBKMO5:b9wk2fd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c979d7cbdb0c2c2376fb70573fe8865818c85b82/pkg/ccl/changefeedccl/avro.go#L510)):*
> ```Go
> 
> func nativeFromHLC(ts hlc.Timestamp) interface{} {
> 	return tree.TimestampToDecimal(ts).Decimal.String()
> ```

I think hlc.Timestamp now has a "AsOfSystemTime" method that can be used instead of this; I'm not fully apprised of subtle differences.
___
*[pkg/ccl/changefeedccl/changefeed.go, line 332 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34095#-LWXMU3k74yfO4447kul:-LWXMU3k74yfO4447kum:br6o115) ([raw file](https://github.com/cockroachdb/cockroach/blob/c979d7cbdb0c2c2376fb70573fe8865818c85b82/pkg/ccl/changefeedccl/changefeed.go#L332)):*
> ```Go
> func emitResolvedTimestamp(
> 	ctx context.Context,
> 	details jobspb.ChangefeedDetails,
> ```

Is this parameter being used?


<!-- Sent from Reviewable.io -->

mrtracy reviewed (commented):
LGTM, just a couple of quick nits/questions.  I will admit that this is still a bit hard to function, but the layers-upon-layers of metadata may just be making this fundamentally impenetrable.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34095#-:-LWX__3H9odgWvIebnoA:bj4recc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @mrtracy)



<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
Thanks for the review, Matt! Good working with ya!

> I will admit that this is still a bit hard to function, but the layers-upon-layers of metadata may just be making this fundamentally impenetrable.

Totally agree. I blame this on the avro library actually. I think if it was structured as more of a builder style pattern, given some codec, this code would all be much more straightforward, and there'd be quite a few less allocs. C'est la vie.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34095#-:-LWXvGad3XBu6PO4wOZf:b1or2yz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mrtracy)

___
*[pkg/ccl/changefeedccl/avro.go, line 510 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34095#-LWXYI6D0S_Ae2uBKMO4:-LWXusd_0ALorxNRrBpc:bcyd8zn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c979d7cbdb0c2c2376fb70573fe8865818c85b82/pkg/ccl/changefeedccl/avro.go#L510)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

I think hlc.Timestamp now has a "AsOfSystemTime" method that can be used instead of this; I'm not fully apprised of subtle differences.
</blockquote></details>

Nice! Done
___
*[pkg/ccl/changefeedccl/changefeed.go, line 332 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34095#-LWXMU3k74yfO4447kul:-LWXv3t88nroJVJH90GS:bxu5w03) ([raw file](https://github.com/cockroachdb/cockroach/blob/c979d7cbdb0c2c2376fb70573fe8865818c85b82/pkg/ccl/changefeedccl/changefeed.go#L332)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Is this parameter being used?
</blockquote></details>

Whoops, artifact of a path I didn't end up going down. Done


<!-- Sent from Reviewable.io -->
